lib/collection/src/shards/resharding/driver.rs

```diff
@@ -559,15 +568,41 @@ async fn stage_replicate(
 /// Stage 4: commit new hashring
 ///
 /// Check whether the new hashring still needs to be committed.
-fn completed_commit_hashring() -> bool {
-    todo!()
+fn completed_commit_hashring(state: &PersistedState) -> bool {
+    state.read().all_peers_reached(Stage::S4_CommitHashringEnd)
 }
 
 /// Stage 4: commit new hashring
 ///
 /// Do commit the new hashring.
-fn stage_commit_hashring() -> CollectionResult<()> {
-    todo!()
+async fn stage_commit_hashring(
+    reshard_key: &ReshardKey,
+    state: &PersistedState,
+    consensus: &dyn ShardTransferConsensus,
+    channel_service: &ChannelService,
+    collection_id: &CollectionId,
+) -> CollectionResult<()> {
+    state.write(|data| {
+        data.bump_all_peers_to(Stage::S4_CommitHashringStart);
+    })?;
+
+    // Commit read hashring and sync cluster
+    consensus
+        .commit_read_hashring_confirm_and_retry(collection_id, reshard_key)
+        .await?;
+    await_consensus_sync(consensus, channel_service).await;
+
+    // Commit write hashring and sync cluster
+    consensus
+        .commit_write_hashring_confirm_and_retry(collection_id, reshard_key)
+        .await?;
+    await_consensus_sync(consensus, channel_service).await;
```

Comment:
The important part is here. Here we commit the read and write hashring in two separate steps. In both cases we sync all peers in the cluster right after to ensure everybody reached the same state.

The other changes in this PR is mostly to route the consensus messages and to confirm/retry.